File: modules/navigation-module/ios/NavigationPresenter.swift
//
//  NavigationPresenter.swift
//  
//
//  Created by Damien L Thompson on 2026-09-02.
//
import CoreLocation
import MapboxDirections
import MapboxNavigationUIKit

@MainActor
protocol NavigationModuleViewProtocol: AnyObject {
    func render(_ state: ViewState<NavigationViewController>)
    func handleArrival(latitude: Double, longitude: Double)
    func handleCancel(reason: String)
}

@MainActor
protocol NavigationPresenterProtocol: AnyObject {
    func configure(origin: CoordinatesRecord, destination: CoordinatesRecord, mode: NavigationMode)
    func teardown()
}

/// Dedupes redundant `configure` calls — Expo re-invokes
/// `OnViewDidUpdateProps` on every prop diff, not just on real changes.
private struct NavigationRequest: Equatable {
    let origin: CoordinatesRecord
    let destination: CoordinatesRecord
    let mode: NavigationMode

    static func == (lhs: Self, rhs: Self) -> Bool {
        lhs.origin.latitude == rhs.origin.latitude &&
        lhs.origin.longitude == rhs.origin.longitude &&
        lhs.destination.latitude == rhs.destination.latitude &&
        lhs.destination.longitude == rhs.destination.longitude &&
        lhs.mode.rawValue == rhs.mode.rawValue
    }
}

@MainActor
final class NavigationPresenter: NSObject, NavigationPresenterProtocol {

    private weak var view: NavigationModuleViewProtocol?
    private let service: NavigationServiceProtocol
    private let builder: NavigationBuilderProtocol

    private var navigationViewController: NavigationViewController?
    private var currentRequest: NavigationRequest?
    private var buildTask: Task<Void, Never>?

    init(
        view: NavigationModuleViewProtocol,
        service: NavigationServiceProtocol,
        builder: NavigationBuilderProtocol
    ) {
        self.view = view
        self.service = service
        self.builder = builder
    }

    func configure(origin: CoordinatesRecord, destination: CoordinatesRecord, mode: NavigationMode) {
        let request = NavigationRequest(origin: origin, destination: destination, mode: mode)
        guard request != currentRequest else { return }
        currentRequest = request

        buildTask?.cancel()
        view?.render(.loading)

        buildTask = Task {
            do {
                let session = try await service.prepareNavigationSession(
                    origin: origin.coordinate,
                    destination: destination.coordinate,
                    profileIdentifier: mode.profileIdentifier,
                    simulateRoute: false
                )
                guard !Task.isCancelled else { return }

                let viewController = builder.buildViewController(from: session)
                viewController.delegate = self

                navigationViewController = viewController
                view?.render(.success(viewController))
            } catch {
                guard !Task.isCancelled else { return }
                view?.render(.error(message: error.localizedDescription))
            }
        }
    }

    func teardown() {
        buildTask?.cancel()

        if let viewController = navigationViewController {
            viewController.delegate = nil
            viewController.navigationMapView?.removeRoutes()
            viewController.navigationMapView?.navigationCamera.stop()
            navigationViewController = nil
        }

        view?.render(.idle)
        currentRequest = nil

        service.resetNavigationSession()
    }
}

// MARK: - NavigationViewControllerDelegate
extension NavigationPresenter: NavigationViewControllerDelegate {
    func navigationViewControllerDidDismiss(
        _ navigationViewController: NavigationViewController,
        byCanceling canceled: Bool
    ) {
        view?.handleCancel(reason: canceled ? "user" : "system")
        teardown()
    }

    func navigationViewController(
        _ navigationViewController: NavigationViewController,
        didArriveAt waypoint: Waypoint
    ) {
        view?.handleArrival(
            latitude: waypoint.coordinate.latitude,
            longitude: waypoint.coordinate.longitude
        )
    }
}

extension CoordinatesRecord {
    fileprivate var coordinate: CLLocationCoordinate2D {
        CLLocationCoordinate2D(latitude: latitude, longitude: longitude)
    }
}

File: modules/navigation-module/android/src/main/java/expo/modules/navigationmodule/NavigationPresenter.kt
package expo.modules.navigationmodule

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import androidx.core.content.ContextCompat
import com.mapbox.geojson.Point
import expo.modules.navigationmodule.ui.NavigationExperienceHost
import java.lang.ref.WeakReference
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch

interface NavigationModuleViewProtocol {
  fun render(state: ViewState<NavigationExperienceHost>)
  fun handleArrival(latitude: Double, longitude: Double)
  fun handleCancel(reason: String)
}

interface NavigationPresenterProtocol {
  fun configure(origin: CoordinatesRecord, destination: CoordinatesRecord, mode: NavigationMode)
  fun teardown()
}

private data class NavigationRequest(
  val origin: CoordinatesRecord,
  val destination: CoordinatesRecord,
  val mode: NavigationMode,
)

class NavigationPresenter(
  view: NavigationModuleViewProtocol,
  private val context: Context,
  private val service: NavigationServiceProtocol,
  private val builder: NavigationBuilderProtocol,
) : NavigationPresenterProtocol {

  private val viewRef = WeakReference(view)
  private val presenterScope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)

  private var navigationHost: NavigationExperienceHost? = null
  private var currentRequest: NavigationRequest? = null
  private var buildJob: Job? = null

  override fun configure(
    origin: CoordinatesRecord,
    destination: CoordinatesRecord,
    mode: NavigationMode,
  ) {
    val request = NavigationRequest(origin, destination, mode)
    if (request == currentRequest) {
      return
    }
    currentRequest = request

    buildJob?.cancel()
    navigationHost?.unbind()
    navigationHost = null
    viewRef.get()?.render(ViewState.Loading)

    buildJob = presenterScope.launch {
      if (!hasLocationPermission()) {
        viewRef.get()?.render(ViewState.Error("Location permission required"))
        return@launch
      }

      try {
        val session = service.prepareNavigationSession(
          origin = origin.toPoint(),
          destination = destination.toPoint(),
          profile = mode.directionsProfile,
          simulateRoute = false,
        )

        if (!isActive) {
          return@launch
        }

        val host = builder.build(context, session).apply {
          onArrived = { _, _ ->
            viewRef.get()?.handleArrival(destination.latitude, destination.longitude)
          }
          onCancel = { reason ->
            viewRef.get()?.handleCancel(reason)
            teardown()
          }
        }

        navigationHost = host
        viewRef.get()?.render(ViewState.Success(host))
      } catch (error: Throwable) {
        if (!isActive) {
          return@launch
        }
        viewRef.get()?.render(ViewState.Error(error.message ?: "Navigation failed"))
      }
    }
  }

  override fun teardown() {
    buildJob?.cancel()
    buildJob = null

    navigationHost?.unbind()
    navigationHost = null

    viewRef.get()?.render(ViewState.Idle)
    currentRequest = null

    service.resetNavigationSession()
    presenterScope.cancel()
  }

  private fun hasLocationPermission(): Boolean {
    return ContextCompat.checkSelfPermission(
      context,
      Manifest.permission.ACCESS_FINE_LOCATION,
    ) == PackageManager.PERMISSION_GRANTED
  }
}

private fun CoordinatesRecord.toPoint(): Point =
  Point.fromLngLat(longitude, latitude)
